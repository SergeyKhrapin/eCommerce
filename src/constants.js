// General
export const SERVER_URL                       = '../../products.json';
export const INCREASE                         = 'INCREASE';
export const DECREASE                         = 'DECREASE';
export const ONE_PRODUCT_MAX_QUANTITY_IN_CART = 5;
export const ONE_PRODUCT_MAX_QUANTITY_ALERT   = `Max limit per product in the cart - ${ONE_PRODUCT_MAX_QUANTITY_IN_CART} items.`;
export const TOTAL_MAX_QUANTITY_IN_CART       = 20;
export const TOTAL_MAX_QUANTITY_ALERT         = `Max products limit in the cart - ${TOTAL_MAX_QUANTITY_IN_CART} items.`;
export const PRODUCT_SUCCESSFULLY_ADDED_ALERT = 'Product is successfully added to the cart.';
export const ALERT_HIDE_DELAY                 = 5000;
export const CURRENCY                         = '$';

// Category page
export const PRUDUCTS_UNAVAILABLE_MESSAGE     = 'The products are unavailable. Sorry for inconvenience.';
export const VIEW_DETAILS                     = 'view details';
export const ADD_TO_CART                      = 'add to cart';

// Cart page
export const CART_HEADING                     = 'Shopping Cart';
export const CART_EMPTY                       = 'Your cart is empty.'
export const VIEW_CART                        = 'View cart';
export const CART_POPUP                       = 'My cart'
export const PRODUCT                          = 'Product';
export const QUANTITY                         = 'Quantity';
export const TOTAL                            = 'Total';
export const SUBTOTAL                         = 'Subtotal';
export const ACTION                           = 'Action';
export const COLOR                            = 'Color';
export const CART_OVERVIEW                    = 'Cart overview';
export const CONTINUE_SHOPPING                = 'Continue shopping';
export const CHECKOUT                         = 'Checkout';
