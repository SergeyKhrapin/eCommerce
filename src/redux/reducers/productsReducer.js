import { FETCH_PRODUCTS } from "../actionTypes";

export const productsReducer = (state = [], action) => {
    switch (action.type) {
        case FETCH_PRODUCTS:
            return state.length ? state : state.concat(action.payload);
        default:
            return state;
    }
}
