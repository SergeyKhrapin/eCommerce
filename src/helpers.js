import { CURRENCY, ONE_PRODUCT_MAX_QUANTITY_IN_CART } from "./constants";

export const getImageSrc = src => `/media/${src}`;

export const getFormattedPrice = (price, multiplier = 1) => CURRENCY + '' + Number(price * multiplier).toFixed(2);

export const getProductDetails = (products, id) => {
    return products.find(product => parseInt(id) === parseInt(product.id)) || {};
    
};

export const getPrettyTitleURL = title => title.split(' ').join('-');

export const increaseQuantity = (quantity, handler) => {
    if (quantity < ONE_PRODUCT_MAX_QUANTITY_IN_CART) {
        handler();
    }
};

export const decreaseQuantity = (quantity, handler) => {
    if (quantity > 1) {
        handler();
    }
};
