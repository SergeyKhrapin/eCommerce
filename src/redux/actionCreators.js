import { SERVER_URL, ALERT_HIDE_DELAY } from "../constants";
import { FETCH_PRODUCTS, ADD_PRODUCT_TO_CART, DECREASE_PRODUCT_QUANTITY_IN_CART, REMOVE_PRODUCT_FROM_CART, HIDE_ALERT, OPEN_CART_POPUP, CLOSE_CART_POPUP } from "./actionTypes";

export const fetchProducts = () => {
    return async dispatch => {
        const response     = await fetch(SERVER_URL);
        const responseJSON = await response.json();

        dispatch({
            type: FETCH_PRODUCTS,
            payload: responseJSON
        });
    }
};

export const addProductToCart = (product, quantity = 1) => {
    // Default value quantity = 1 is used when a product is added to cart from PLP or its quantity increased from Cart page
    return dispatch => {
        dispatch({
            type: ADD_PRODUCT_TO_CART,
            payload: { product, quantity }
        });

        setTimeout(() => {
            dispatch(hideAlert());
        }, ALERT_HIDE_DELAY);
    };
};

export const decreaseProductQuantityInCart = (product, quantity = 1) => {
    return {
        type: DECREASE_PRODUCT_QUANTITY_IN_CART,
        payload: { product, quantity }
    }
};

export const removeProductFromCart = (product, quantity) => {
    return {
        type: REMOVE_PRODUCT_FROM_CART,
        payload: { product, quantity }
    }
};

export const hideAlert = () => {
    return {
        type: HIDE_ALERT
    }
};

export const openCartPopup = () => {
    return {
        type: OPEN_CART_POPUP
    }
};

export const closeCartPopup = () => {
    return {
        type: CLOSE_CART_POPUP
    }
};
