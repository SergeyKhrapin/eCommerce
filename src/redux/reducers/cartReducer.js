import { ADD_PRODUCT_TO_CART, CLOSE_CART_POPUP, DECREASE_PRODUCT_QUANTITY_IN_CART, HIDE_ALERT, OPEN_CART_POPUP, REMOVE_PRODUCT_FROM_CART, SHOW_ALERT } from "../actionTypes";
import { ONE_PRODUCT_MAX_QUANTITY_IN_CART, ONE_PRODUCT_MAX_QUANTITY_ALERT, TOTAL_MAX_QUANTITY_IN_CART, TOTAL_MAX_QUANTITY_ALERT, PRODUCT_SUCCESSFULLY_ADDED_ALERT } from '../../constants';

const initialState = {
    products: {},
    totalQuantity: 0,
    totalPrice: 0,
    alert: null,
    openPopup: null
};

export const cartReducer = (state = initialState, action) => {
    const { products: productsInCart,
          totalQuantity: productsQuantityInCart,
          totalPrice } = state;
    const { type, payload } = action;
    const { product, quantity } = payload ?? {};
    
    switch (type) {
        case ADD_PRODUCT_TO_CART:
            // If the product is already in the cart - increase its quantity
            // If there is no yet this product in the cart:
            // ~~undefined?.quantity + quantity >>> ~~undefined + quantity >>> 0 + quantity
            const increasedQuantity = ~~productsInCart[product.id]?.quantity + quantity;

            // Max limit all products in the cart reached
            if (productsQuantityInCart + quantity > TOTAL_MAX_QUANTITY_IN_CART) {
                return {
                    ...state,
                    alert: TOTAL_MAX_QUANTITY_ALERT
                };
            }
            
            // Max limit per one product in the cart reached
            if (increasedQuantity > ONE_PRODUCT_MAX_QUANTITY_IN_CART) {
                return {
                    ...state,
                    alert: ONE_PRODUCT_MAX_QUANTITY_ALERT
                };
            }

            return {
                ...state,
                products: {
                    ...productsInCart,
                    [product.id]: {product, quantity: increasedQuantity}
                },
                totalQuantity: productsQuantityInCart + quantity,
                totalPrice: totalPrice + product.price * quantity,
                alert: PRODUCT_SUCCESSFULLY_ADDED_ALERT
            };

        case DECREASE_PRODUCT_QUANTITY_IN_CART:
            return {
                ...state,
                products: {
                    ...productsInCart,
                    [product.id]: {
                        product,
                        quantity: state.products[product.id].quantity - 1
                    }
                },
                totalQuantity: state.totalQuantity - 1,
                totalPrice: totalPrice - product.price
            };

        case REMOVE_PRODUCT_FROM_CART:
            // Remove a product without mutation
            const { [product.id]: omit, ...updatedProductsInCart } = productsInCart;
            return {
                ...state,
                products: updatedProductsInCart,
                totalQuantity: state.totalQuantity - quantity,
                totalPrice: totalPrice - product.price * quantity
            };

        case HIDE_ALERT:
            return {
                ...state,
                alert: false
            };

        case OPEN_CART_POPUP:
            return {
                ...state,
                openPopup: true
            }

        case CLOSE_CART_POPUP:
            return {
                ...state,
                openPopup: false
            }

        default:
            return state;
    }
};
