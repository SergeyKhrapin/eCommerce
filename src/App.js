import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { Route, Link } from 'react-router-dom';
import { fetchProducts } from './redux/actionCreators';
import logo from './logo.png';
import Category from "./category/Category";
import Cart from "./cart/Cart";
import Product from "./product/Product";
import CartPopup from './cartPopup/CartPopup';
import styles from "./app.module.css";

const App = ({ openPopup, fetchProducts }) => {
    useEffect(fetchProducts, [])

    return (
        <div className={`${styles.app} ${openPopup ? styles.appCartPopupOpened : ''}`} data-testid="app-container">
            <header className={styles.header}>
                <div className="container">
                    <div className="row">
                        <div className="col-2">
                            <img src={logo} className={styles.logo} alt="logo" />
                        </div>
                        <div className="col-8">
                            <nav className={styles.nav}>
                                <ul className={styles.list}>
                                    <li><Link to="/">HOME</Link></li>
                                    <li><a href="#">SHOP</a></li>
                                    <li><a href="#">JOURNAL</a></li>
                                    <li><a href="#">MORE</a></li>
                                </ul>
                            </nav>
                        </div>
                        <div className="col-2">
                            <CartPopup />
                        </div>
                    </div>
                </div>
            </header>

            <Route exact path="/" component={Category} />
            <Route path="/cart" component={Cart} />
            <Route path="/product/:id/:title" component={Product} />
        </div>
    );
}

const mapStateToProps = state => {
    return {
        openPopup: state.cart.openPopup
    };
};

const mapDispatchToProps = {
    fetchProducts
}

export default connect(mapStateToProps, mapDispatchToProps)(App)
