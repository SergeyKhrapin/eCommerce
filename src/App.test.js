import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import App from './App';
import * as helpers from './helpers';
import * as actionCreators from './redux/actionCreators';
import * as actionTypes from './redux/actionTypes';
import * as reducer from './redux/reducers/cartReducer';
import * as constant from './constants';
import puppeteer from 'puppeteer';

describe('Helper: ', () => {
    it('getFormattedPrice should transform number 40 to $40.00', () => {
        expect(helpers.getFormattedPrice(40)).toBe('$40.00');
    });

    it('getPrettyTitleURL should replace spaces with dashes', () => {
        expect(helpers.getPrettyTitleURL('Blue Stripe Stoneware Plate')).toBe('Blue-Stripe-Stoneware-Plate');
    });

    it('getProductDetails should return a product or empty object', () => {
        expect(helpers.getProductDetails([])).toBeDefined();
    });
});

describe('Action creator:', () => {
    const product = { id: 1000003, price: 40, title: 'MASHIKO' };
    const action = {
        type: actionTypes.DECREASE_PRODUCT_QUANTITY_IN_CART,
        payload: { product, quantity: 1 }
    }

    it('should return an action object', () => {
        expect(actionCreators.decreaseProductQuantityInCart(product)).toStrictEqual(action);
    });

    it('should return an object with type field', () => {
        expect(actionCreators.decreaseProductQuantityInCart()).toHaveProperty('type');
        expect(actionCreators.removeProductFromCart()).toHaveProperty('type');
        expect(actionCreators.hideAlert()).toHaveProperty('type');
        expect(actionCreators.openCartPopup()).toHaveProperty('type');
        expect(actionCreators.closeCartPopup()).toHaveProperty('type');
    });
});

describe('Reducer:', () => {
    it('should always return a state object', () => {
        const initialStateMock = {
            products: {},
            totalQuantity: 0,
            totalPrice: 0,
            alert: null,
            openPopup: null
        };
        const actionMock = { type: 'SOME_NON_EXISTENT_TYPE' };
        expect(reducer.cartReducer(initialStateMock, actionMock)).toEqual(initialStateMock);
    });
});

describe('Add product to the cart:', () => {
    let state, action;

    beforeEach(() => {
        state = {
            products: {},
            totalQuantity: 0,
            totalPrice: 0,
            alert: null,
            openPopup: null
        };
        action = {
            type: actionTypes.ADD_PRODUCT_TO_CART,
            payload: {
                product: { id: 1000003, price: 40, title: 'MASHIKO' },
                quantity: 0
            }
        };
    });

    it('if total products quantity in cart will exceed max limit - product is not added and alert is displayed', () => {
        state.totalQuantity = 19;
        action.payload.quantity = 2;
        const updatedState = reducer.cartReducer(state, action);
        expect(updatedState.totalQuantity).toBeLessThanOrEqual(constant.TOTAL_MAX_QUANTITY_IN_CART);
        expect(updatedState.alert).toBe(constant.TOTAL_MAX_QUANTITY_ALERT);
    });

    it('if product quantity in cart will exceed max limit per one product - product is not added and alert is displayed', () => {
        action.payload.quantity = 6;
        const updatedState = reducer.cartReducer(state, action);
        expect(updatedState.products['1000003']).not.toBeDefined();
        expect(updatedState.alert).toBe(constant.ONE_PRODUCT_MAX_QUANTITY_ALERT);
    });
});

describe('Remove product from the cart should lead to:', () => {
    let state, action, updatedState;

    beforeEach(() => {
        state = {
            products: {
                1000001: {
                    product: { id: 1000001, price: 50, title: "Heme" },
                    quantity: 2
                },
                1000003: {
                    product: { id: 1000003, price: 40, title: 'MASHIKO' },
                    quantity: 3
                }
            },
            totalQuantity: 5,
            totalPrice: 220,
            alert: null,
            openPopup: null
        };

        action = {
            type: actionTypes.REMOVE_PRODUCT_FROM_CART,
            payload: {
                product: { id: 1000003, price: 40, title: 'MASHIKO' },
                quantity: 3
            }
        };

        updatedState = reducer.cartReducer(state, action);
    });

    it('remove product id field from cart products object', () => {
        expect(updatedState.products).not.toHaveProperty([action.payload.product.id]);
    });

    it('reduce total products quantity by quantity of removed product', () => {
        expect(updatedState.totalQuantity).toEqual(state.totalQuantity - action.payload.quantity);
    });

    it('reduce total price by total price of removed product', () => {
        expect(updatedState.totalPrice).toEqual(state.totalPrice - action.payload.product.price * action.payload.quantity);
    });
});

describe('Decrease a product quantity in the cart should lead to:', () => {
    let state, action, updatedState;

    beforeEach(() => {
        state = {
            products: {
                1000001: {
                    product: { id: 1000001, price: 50, title: "Heme" },
                    quantity: 2
                },
                1000003: {
                    product: { id: 1000003, price: 40, title: 'MASHIKO' },
                    quantity: 3
                }
            },
            totalQuantity: 5,
            totalPrice: 220,
            alert: null,
            openPopup: null
        };

        action = {
            type: actionTypes.DECREASE_PRODUCT_QUANTITY_IN_CART,
            payload: {
                product: { id: 1000003, price: 40, title: 'MASHIKO' },
                quantity: 3
            }
        };

        updatedState = reducer.cartReducer(state, action);
    });

    it('decrease a product quantity by 1', () => {
        expect(updatedState.products['1000003'].quantity).toEqual(2);
    });

    it('decrease total products quantity by 1', () => {
        expect(updatedState.totalQuantity).toEqual(4);
    });

    it('decrease total price by product price', () => {
        expect(updatedState.totalPrice).toEqual(220 - 40);
    });
});

// describe('App component renders:', () => {
//     it('correctly without products', () => {
//         const mockStore = configureStore();
//         const store = mockStore({
//             allProducts: [],
//             cart: {
//                 products: {
//                     1000001: {
//                         product: { id: 1000001, price: 50, title: "Heme" },
//                         quantity: 2
//                     },
//                     1000003: {
//                         product: { id: 1000003, price: 40, title: 'MASHIKO' },
//                         quantity: 3
//                     }
//                 },
//                 totalQuantity: 5,
//                 totalPrice: 220,
//                 alert: null,
//                 openPopup: null
//             }
//         });

//         const component = renderer.create(
//             <Provider store={store}>
//                 <BrowserRouter>
//                     <App />
//                 </BrowserRouter>
//             </Provider>
//         );

//         let tree = component.toJSON();
//         expect(tree).toMatchSnapshot();
//     });
// });

describe('Fetch products:', () => {
    const mockProducts = [
        { id: 1000001, price: 40, title: 'MASHIKO' },
        { id: 1000002, price: 60, title: 'HEME' },
    ];

    const action = {
        type: actionTypes.FETCH_PRODUCTS,
        payload: { allProducts: mockProducts }
    };

    const middlewares = [thunk];
    const mockStore = configureStore(middlewares);

    it('should return an action object with products', async () => {
        window.fetch = jest.fn().mockImplementation(() => ({
            status: 200,
            json: () => new Promise((resolve, reject) => {
                resolve({
                    allProducts: mockProducts
                })
            })
        }))

        const store = mockStore({});

        await store.dispatch(actionCreators.fetchProducts());
        const actions = store.getActions();

        expect(actions[0]).toEqual(action);
    });
});

// describe('Add product to cart:', () => {
//     let store, state, product, action;

//     beforeEach(() => {
//         const middlewares = [thunk];
//         const mockStore = configureStore(middlewares);
//         store = mockStore({});

//         state = {
//             products: {},
//             totalQuantity: 0,
//             totalPrice: 0,
//             alert: null,
//             openPopup: null
//         };

//         product = { id: '1000001', price: 40, title: 'MASHIKO' };

//         action = {
//             type: actionTypes.ADD_PRODUCT_TO_CART,
//             payload: {
//                 product: product,
//                 quantity: 1
//             }
//         };
//     });

//     it('call addProductToCart and setTimeout inside it', () => {
//         jest.useFakeTimers();
//         store.dispatch(actionCreators.addProductToCart(product, 1));
//         expect(setTimeout).toHaveBeenCalled();
//         expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), constant.ALERT_HIDE_DELAY);
//     });

//     it('should return an action object with that product', async () => {
//         const spy = jest.spyOn(actionCreators, 'addProductToCart');
//         await store.dispatch(actionCreators.addProductToCart(product, 1));
//         const actions = store.getActions();

//         expect(actions[0]).toEqual(action);
//         expect(spy).toHaveBeenCalled();
//     });
// });

// describe('When cart popup is opened:', () => {
//     beforeEach(() => {
//         const mockStore = configureStore();

//         const store = mockStore({
//             allProducts: [],
//             cart: {
//                 products: {},
//                 openPopup: true
//             }
//         });

//         render(
//             <Provider store={store}>
//                 <BrowserRouter>
//                     <App />
//                 </BrowserRouter>
//             </Provider>
//         );
//     });

//     it('App container has a corresponding class', () => {
//         const app = screen.getByTestId('app-container');
//         expect(app).toHaveClass('cartPopup-opened');
//     });

//     it('cart popup container and overlay are displayed in DOM', () => {
//         const cartPopup = screen.getByTestId('cart-popup');
//         const cartPopupClose = screen.getByTestId('cart-popup-close');
//         expect(cartPopup).toBeInTheDocument();
//         expect(cartPopupClose).toBeInTheDocument();
//     });

//     it('open cart popup link has aria-expanded=true', () => {
//         const cartPopupOpenLink = screen.getByTestId('cart-popup-open-link');
//         expect(cartPopupOpenLink).toHaveAttribute('aria-expanded', 'true');
//     });
// });

/* ================
    e2e tests
=================== */

test('Homepage is loaded successfully', async () => {
  const browser = await puppeteer.launch({
    headless: false,
    defaultViewport: null,
    args: ['--window-size=1530,800'],
  });
  const page = await browser.newPage();
  await page.goto('http://localhost:3000/');

	// add 2 products to cart
	await page.hover('#productTile-1000001');
	await page.click('#product-1000001');
	await page.waitForTimeout(1000);
	await page.hover('#productTile-1000002');
	await page.click('#product-1000002');
	
	// open cart popup
	await page.evaluateHandle(() => {
		document.querySelector('[data-testid="cart-popup-open-link"]').click();
	})

	// check if cart popup is really opened
	const isPopupOpened = await page.$$eval('[data-testid="cart-popup"]', items => {
		// return item; Known issue - puppeteer returns undefined!
		return items.length;
	})
	expect(isPopupOpened).toEqual(1);

	// check the amount of items in the cart
	const itemsInCart = await page.$$eval('[data-testid="cart-popup-item"]', items => items.length)
	// TODO: this fails sometimes when tests run for the first time!
	expect(itemsInCart).toEqual(2);

	// await browser.close();
}, 100000)
