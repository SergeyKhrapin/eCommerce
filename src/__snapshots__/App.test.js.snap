// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders: correctly without products 1`] = `
<div
  className="app "
  data-testid="app-container"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="row"
      >
        <div
          className="col-2"
        >
          <img
            alt="logo"
            className="logo"
            src="logo.png"
          />
        </div>
        <div
          className="col-8"
        >
          <nav
            className="nav"
          >
            <ul
              className="list"
            >
              <li>
                <a
                  href="/"
                  onClick={[Function]}
                >
                  HOME
                </a>
              </li>
              <li>
                <a
                  href="#"
                >
                  SHOP
                </a>
              </li>
              <li>
                <a
                  href="#"
                >
                  JOURNAL
                </a>
              </li>
              <li>
                <a
                  href="#"
                >
                  MORE
                </a>
              </li>
            </ul>
          </nav>
        </div>
        <div
          className="col-2"
        >
          <div
            className="cartPopupOpen"
          >
            <span
              aria-expanded="false"
              className="cartPopupOpenLink"
              data-testid="cart-popup-open-link"
              onClick={[Function]}
            >
              My cart (5)
            </span>
          </div>
        </div>
      </div>
    </div>
  </header>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="categoryBanner"
        >
          <img
            alt="Home page banner"
            className="categoryBannerImage"
            src="/media/plates-header.jpg"
          />
          <div
            className="categoryBannerContent"
          >
            <h1>
              Plates
            </h1>
            <p>
              Lorem ipsum dolor, sit amet consectetur adipisicing elit. Ipsum eveniet rem ad culpa quidem incidunt laborum accusamus eius earum animi aliquid corporis inventore ex, dignissimos consequuntur expedita, vitae distinctio odit.
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row mt-5"
    >
      <h1>
        The products are unavailable. Sorry for inconvenience.
      </h1>
    </div>
  </div>
</div>
`;
